BaseDto to support Save and Restore state

GioviQ opened the pull request.

enkodellc commented on src/BlazorBoilerplate.Server/Startup.cs:
```diff
@@ -342,12 +342,6 @@ void DbContextOptionsBuilder(DbContextOptionsBuilder builder)
                 };
             });
 
-            services.AddResponseCompression(opts =>
```
Is there a reason to remove? I think I tried removing it from a previous PR and I was getting errors. I will research.

enkodellc commented:
The PR looks great except for the Response Compression removal in Startup. Please let me know the reason to remove. I think it was asked to be removed in the docs but I also then got runtime errors when I removed it. I will review and test now but this was my initial code review of the PR. 

Really appreciate it @GioviQ!

MarkStega commented:
I removed it in my 'real' solution and everything continues to function.

GioviQ commented:
> The PR looks great except for the Response Compression removal in Startup. Please let me know the reason to remove. I think it was asked to be removed in the docs but I also then got runtime errors when I removed it. I will review and test now but this was my initial code review of the PR.
> 
> Really appreciate it @GioviQ!

@enkodellc I removed response compression as stated in https://devblogs.microsoft.com/aspnet/blazor-webassembly-3-2-0-preview-2-release-now-available/

Kronos11 commented:
I'm really unsure about UltraMapper and why you would need this in combination with AutoMapper?

GioviQ commented:
> I'm really unsure about UltraMapper and why you would need this in combination with AutoMapper?

@Kronos11 I did not want to include AutoMapper in https://github.com/enkodellc/blazorboilerplate/blob/restructure/src/BlazorBoilerplate.Shared/BlazorBoilerplate.Shared.csproj

I prefer UltraMapper (https://github.com/GioviQ/UltraMapper) because it is light and simple (no configuration needed).
It was easy to me implement Memento pattern to save e restore a DTO.

soomon commented:
@GioviQ as I did not find an issue for this: could you please explain to me what saving and restoring a dto is all about?

GioviQ commented:
@soomon I implement Memento pattern because the issue https://github.com/enkodellc/blazorboilerplate/issues/217

soomon commented:
Thanks mate. I checked the code of the user admin page and I do not understand why that issue exists at all. Can you please point out to me where/why the data modified in the user edit dialog gets pushed back to the table data?

GioviQ commented:
Edit an user. The edit dialog opens. Change email, user name, etc. Click Cancel button.
Dialog closes. The data table shows the user fields changed because UserInfoDto changed. You have to reload to see the true values of db.
This happen before my PR @soomon

soomon commented:
I am able to reproduce the issue. 

Here is what I do not understand:
Isn't the IserInfoDto user completely separate from the List<UserInfoDto> users?
Where is the connection? 

Can you please let me know the line of code (before your PR) that causes the issue?
I fail to see where the binding gets created that pushes the change inside "user" into "users"

GioviQ commented:
https://github.com/enkodellc/blazorboilerplate/blob/master/src/BlazorBoilerplate.CommonUI/Pages/Admin/Users.razor
(row 198) list of UserInfoDto are binded in row 21
every table row has UserRow as a context which is one UserInfoDto of the list.

When you open the edit dialog you are working on one of this row retrieved in
row 209, so when you change it the table row reflect the change. This is how binding works.

soomon commented:
Thanks mate.
`Line 212 reads: user = users.Where(x => x.UserId == userId).SingleOrDefault();  // load the user information into the temp user object for worry free editing`
As far as I understand it, this is copying the information, not binding it? 
The edit Dialog then using the copied object, not the one in the users list.
Or is that wrong?

If I do this:
```
        tempuser = new UserInfoDto();
        user = users.Where(x => x.UserId == userId).SingleOrDefault();  // load the user information into the temp user object for worry free editing

        user.DisableTenantFilter = tempuser.DisableTenantFilter;
        user.Email = tempuser.Email;
        user.ExposedClaims = tempuser.ExposedClaims;
        user.FirstName = tempuser.FirstName;
        user.IsAuthenticated = tempuser.IsAuthenticated;
        user.LastName = tempuser.LastName;
        user.Roles = tempuser.Roles;
        user.TenantId = tempuser.TenantId;
        user.UserId = tempuser.UserId;
        user.UserName = tempuser.UserName;
```

The error goes away. Of course, the updated object needs to replace the one in the list.
So if all we need to do is copying the object yet again, can't we use automapper to do so?

GioviQ commented:
I think AutoMapper can do it, try it, but I said before UltraMapper seems simpler and lighter.
